sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1041,11 +1042,26 @@ public long getSplitBacklogBytes() {
 
     @Override
     public void close() throws IOException {
-      closed = true;
-      availableRecordsQueue.poll(); // drain unread batch, this unblocks consumer thread.
-      consumer.wakeup();
+      closed.set(true);
       consumerPollThread.shutdown();
       offsetFetcherThread.shutdown();
+      while (true) {
+        // drain unread batch, this unblocks consumer thread. trying this in a loop to
+        // handle a small race where poll thread might try to enqueue after we drain.
+        consumer.wakeup();
+        availableRecordsQueue.poll();
+        try {
+          if (consumerPollThread.awaitTermination(10, TimeUnit.SECONDS)
+              && offsetFetcherThread.awaitTermination(10, TimeUnit.SECONDS)) {
+            break; // done
+          }
+        } catch (InterruptedException e) {
```

Comment:
I set the flag in two places where we are interrupted on thread not in our control.